File: backend/app/api/v1/schemas/campanhas.py
from datetime import date
from uuid import UUID

from pydantic import BaseModel, Field

from shared.enums import Perfil, StatusCampanha


class CreateCampaignIn(BaseModel):
    nome: str
    tipo: str
    descricao: str = ""
    publico: list[Perfil] = Field(min_length=1)
    questionario_id: UUID
    inicio: date
    fim: date


class CampaignOut(BaseModel):
    id: UUID
    nome: str
    tipo: str
    descricao: str
    inicio: date
    fim: date
    participacao: float
    respostas: int
    publico: str
    publico_perfis: list[Perfil]
    questionario: str
    questionario_id: UUID
    status: StatusCampanha
    categoria: str

File: frontend/src/api/campanhas.ts
import { apiClient } from './client'
import type { Campanha, PerfilParticipante, StatusCampanha } from '../data/mock'
import { isoToBr } from '../utils/date'

interface CampanhaRaw {
  id: string
  nome: string
  tipo: string
  descricao: string
  inicio: string
  fim: string
  participacao: number
  respostas: number
  publico: string
  publico_perfis: PerfilParticipante[]
  questionario: string
  questionario_id: string
  status: StatusCampanha
  categoria: string
}

export interface CriarCampanhaInput {
  nome: string
  tipo: string
  descricao: string
  publico: PerfilParticipante[]
  questionario_id: string
  inicio: string
  fim: string
}

function mapCampanha(raw: CampanhaRaw): Campanha {
  return {
    id: raw.id,
    nome: raw.nome,
    tipo: raw.tipo,
    descricao: raw.descricao,
    inicio: isoToBr(raw.inicio),
    fim: isoToBr(raw.fim),
    participacao: raw.participacao,
    respostas: raw.respostas,
    publico: raw.publico,
    publicoPerfis: raw.publico_perfis,
    questionario: raw.questionario,
    questionarioId: raw.questionario_id,
    status: raw.status,
    categoria: raw.categoria,
  }
}

export async function listCampanhas(signal?: AbortSignal): Promise<Campanha[]> {
  const response = await apiClient.get<CampanhaRaw[]>('/api/v1/campanhas', { signal })
  return response.map(mapCampanha)
}

export async function createCampanha(input: CriarCampanhaInput): Promise<Campanha> {
  const response = await apiClient.post<CampanhaRaw>('/api/v1/campanhas', input)
  return mapCampanha(response)
}
